### ISSUE TITLE
Mapping of ODM 2 variables+units+sampledmedium concepts to ODM 1 variables

### ISSUE THREAD

**Issue by emiliom** [ODM2/WOFpy#15]
Examining the [ODM2 WOFpy test endpoint](http://sis-devel.cloudapp.net:8080/), I've noticed that the treatment of "variables" in a WOF response (which should match the [ODM1 concept of variables](http://his.cuahsi.org/odmdatabases.html), as found in its Variables entity) from an ODM2 database needs refinement. Despite the similarities in names and carrying over a lot of the variable name vocabulary terms from ODM 1 to ODM 2, the two in reality can not be assume to map 1:1.

In [ODM1](http://his.cuahsi.org/images/ODM1_1SchemaDiagram_md.jpg), several concepts were merged into a single "variable": `VariableName`, `SampleMedium`, Units (as `VariableUnitsID`), and others. Each `VariableCode` is a distinct combination of those concepts, and those should be the individual entities found in the [`GetVariables` response, such as the ODM2 WOFpy response](http://sis-devel.cloudapp.net:8080/GetVariableInfo).

But in [ODM2, the scope of the Variables entity is more generic and potentially broader](http://odm2.github.io/ODM2/schemas/ODM2_Current/tables/ODM2Core_Variables.html). Because the [concepts of `SampleMedium` and `Units` (among others) have been moved to be associated with `Results`](http://odm2.github.io/ODM2/schemas/ODM2_Current/diagrams/ODM2Core.html), not with `Variables`, different combinations can arise. A single `VariableCode` may now be associated with multiple `SampledMediumCV` and `UnitsID` values. This is in fact what I did in the Marchantaria use case; for example, the *FPOC* `VariableCode` occurs in combination with two different `UnitsID` (and corresponding units, with physically very different `UnitsTypeCV`). 

You can see an illustration of this ODM2 possibility in [my IPython notebook ODM2REST API demo from August](http://nbviewer.ipython.org/github/BiG-CZ/BiG-CZ-Toolbox/blob/master/ipynotebooks/ODM2RESTdemo_MarchantariaUseCase.ipynb), where I used FPOC VariableCode as one example. See Cell 6, the summary on Cell 15 showing the "Same VariableCode, but distinguished by UnitsTypeCV", and the `ResultValues` table and plot in Cells 14 and 16, respectively. FPOC in mg/L and Mass % are physically different things, as shown in the plot. In contrast, the ODM2 WOFpy implementation is apparently assuming that there's only one `UnitsID` and one `SampledMediumCV` associated with each `VariableCode`, as you can see on the [GetVariableInfo?variable=ODM2WofPy:FPOC response](http://sis-devel.cloudapp.net:8080/GetVariableInfo?variable=ODM2WofPy:FPOC). This assumption gives mangled results, as `GetValues` with `variable=ODM2WofPy:FPOC` returns both types of FPOC variables while incorrectly assigning the assumed, single units of mg/L shown on the `GetVariableInfo?variable=ODM2WofPy:FPOC` response.

It seems to me that the only way to resolve this for a proper WOF/WaterML 1 response that's based on an ODM2 database is to create an on-the-fly equivalent of the ODM1 VariableCode, that is not the same as `Variables.VariableCode`. I'll give a suggestion on a follow-up comment, with PostgreSQL SQL code that I've tested on my Marchantaria use case.

**Comment by emiliom** [comment]
To follow up with SQL code, first I created a view (FYI, as a [materialized view](http://www.postgresql.org/docs/current/static/sql-creatematerializedview.html), for speed) that's probably fairly close to the ODM2REST API concept of "Observations". I've pasted the SQL code for this `observations_mv1` view at the bottom, for reference.

Then, I use the `observations_mv1` view to bring in fields from Variables, Units, and Results, combining these ODM2 concepts (via unique terms or identifiers for each of them) into an analog of the ODM1 VariableCode concept, which below is aliased as the field name `varunitsm_cd`:
```SQL
select
  variablecode || '::' || unitsid || '-' || sampledmediumcv AS varunitsm_cd,
  count(*) AS site_result_count,
  min(begindatetime) AS site_action_begindt,
  max(begindatetime) AS site_action_enddt,
  variablecode, variablenamecv, unitstypecv, unitsabbreviation, sampledmediumcv, 
  variableid, unitsid
from odm2.observations_mv1
group by
  variablecode, variablenamecv, unitstypecv, unitsabbreviation, sampledmediumcv, 
  variableid, unitsid
order by
  variablecode, variablenamecv, unitstypecv, unitsabbreviation, sampledmediumcv, 
  variableid, unitsid;
```
This isn't perfect, but, for example, it does distinguish between the two FPOC occurrences I've described. The two string values of `varunitsm_cd` for FPOC are *"FPOC::1-Liquid aqueous"* and *"FPOC::12-Liquid aqueous"*, where 1 and 12 are the two different `UnitsID` values (FYI, ideally these strings should be sanitized, eg, to remove blank characters).

BTW, note that this query does something else, that I'll comment on in a separate issue: it provides correct datetime ranges for each ODM1-like VariableCode; [currently the ODM2 WOFpy `seriesCatalog` element in `GetSiteInfo` requests is not doing this](http://sis-devel.cloudapp.net:8080/GetSiteInfo?site=ODM2WofPy:march).

This is my "Observations"-like view:
```SQL
CREATE MATERIALIZED VIEW odm2.observations_mv1 AS
SELECT 
  actions.actionid, 
  actions.actiontypecv, actions.begindatetime, actions.enddatetime, 
  actions.actiondescription, 
  actions.methodid, 
  methods.methodtypecv, methods.methodcode, methods.methodname, 
  methods.organizationid, 
  samplingfeatures.samplingfeatureid, 
  samplingfeatures.samplingfeaturetypecv, samplingfeatures.samplingfeaturecode, 
  results.resultid, 
  results.featureactionid, results.resulttypecv, 
  results.taxonomicclassifierid, results.resultdatetime, results.statuscv, 
  results.sampledmediumcv, results.valuecount, 
  variables.variableid, 
  variables.variabletypecv, variables.variablecode, variables.variablenamecv, 
  variables.variabledefinition, variables.speciationcv, 
  units.unitsid, 
  units.unitstypecv, units.unitsabbreviation, units.unitsname
FROM 
  odm2.actions, odm2.results, odm2.featureactions, odm2.methods, 
  odm2.units, odm2.variables, odm2.samplingfeatures
WHERE 
  actions.actionid = featureactions.actionid AND
  actions.methodid = methods.methodid AND
  results.featureactionid = featureactions.featureactionid AND
  results.variableid = variables.variableid AND
  results.unitsid = units.unitsid AND
  samplingfeatures.samplingfeatureid = featureactions.samplingfeatureid
ORDER BY
  actions.begindatetime, samplingfeatures.samplingfeatureid, 
  results.resultid, results.variableid, results.unitsid;
```

**Comment by valentinedwv** [comment]
Thanks.
Choonhan, can we look into how to do this as an object, then see if it fits into the ODM2API.

**Comment by cdesyoun** [comment]
Emilio, it also very useful information. David and I discussed about your dataset for the implementation. Based on these comments, currently we have focused on specimen measurement data use case rather than time series data.
With WaterML1.1, we will release new version of WOFpy services for this datset tomorrow some time.

**Comment by emiliom** [comment]
Thanks, Choonhan. I'm glad the data use case continues to be useful. 

While my data use case is much like the time series data WOF & WaterML 1 are optimized for, @horsburgh was absolutely right that we should focus on time series first. Once you start looking at other specimen-based use cases, you'll find that they can be very diverse and can quickly break the capabilities of WaterML 1. So I recommend you do stay focused on time series data (mainly sensor data, but also my use case!) for the WOFpy application. I think the wide diversity of specimen data can be best explored with other data encodings (like EarthChemXML and WQX) and your ODM2REST service. But getting a fully working WOFpy that is widely deployable, hooks up to ODM2 databases and is WOF 1.1/WaterML 1.1 compliant will be a really big deal in getting ODM2 to start being interoperable with other systems!

My 2 cents, anyway.

**Comment by valentinedwv** [comment]
Waterml 1.1 should work for all data access objects(doa), although others will need a bit of validation.

first in, first out. odm2 wofpy doa is evolving into two data access objects: measurements and time series, with plans for others as we get example databases. This will better allow for agile improvements to the specimens cases.

**Comment by emiliom** [comment]
> Waterml 1.1 should work for all data access objects(doa)

@valentinedwv, are you saying that's now ready and accessible on the test service? If so -- Terrific! And, Issue #2 can be closed.

> first in, first out. odm2 wofpy doa is evolving into two data access objects: measurements and time series, with plans for others as we get example databases. This will better allow for agile improvements to the specimens cases.

Sounds good. Support for Measurements will inevitably have limits, though, b/c of the nature of WOF & WaterML 1. Getting the (sensor-based) Time Series Result type and to a lesser extent the "time-series-like" Measurement Result type in top shape first will be really great.

**Comment by cdesyoun** [comment]
Emilio, I released WOFpy 1.0 and WOFpy 1.1 through other email.

**Comment by emiliom** [comment]
Yes, I saw that. Thanks!!
